Edits up to Deterministic Mappings. (#267)
cfrg/draft-irtf-cfrg-hash-to-curve: draft-irtf-cfrg-hash-to-curve.md

as hashing to an elliptic curve. Prominent examples of cryptosystems that
hash to elliptic curves include Simple Password Exponential Key Exchange
{{J96}}, Password Authenticated Key Exchange {{BMP00}}, Identity-Based
Encryption {{BF01}}, Boneh-Lynn-Shacham signatures {{BLS01}} {{?I-D.irtf-cfrg-bls-signature}},
Verifiable Random Functions {{?I-D.irtf-cfrg-vrf}}, and Oblivious Pseudorandom
Functions {{?I-D.irtf-cfrg-voprf}}.

Unfortunately for implementors, the precise hash function that is suitable for a
given scheme is not necessarily included in the description of the protocol.
Compounding this problem is the need to pick a suitable curve for the specific
protocol. Incorrect choice of hash function or curve can have disastrous

> consequences on the security of a protocol.

Reviewer: ```suggestion
consequences for the security of a protocol.
```

This group has order n, meaning that there are n distinct points.
This document uses additive notation for the elliptic curve group operation.

> For security reasons, protocols using elliptic curve require that the group
> be of prime order. Elliptic curves induce subgroups of prime order.

Reviewer: ```suggestion
For security reasons, protocols using elliptic curves require a group
of prime order. Elliptic curves induce subgroups of prime order.
```

Encodings are closely related to mappings.
Like a mapping, an encoding is a function that outputs a point on an elliptic curve.
In contrast to a mapping, however, the input to an encoding is an arbitrary-length
byte string. Encodings can be deterministic or probabilistic.
Deterministic encodings are preferred for security, because probabilistic
ones are more likely to leak information through side channels. See {{VR20}}
as an example of this type of leakage leading to a security vulnerability.

> This document constructs deterministic encodings by composing a hash function H'

Reviewer: As an alternative, I was going to suggest `Hf`. Maybe slightly easier to read than H'?

Author: That'll work too! Want to make that change in your pass after rebasing?

6. return P
~~~

Instances of these encoding functions are given in {{suites}}, which defines a
list of suites that specify a full set of parameters matching elliptic curves
and algorithms.

## Domain separation requirements {#domain-separation}

All uses of the encoding functions defined in this document MUST include
domain separation ({{term-domain-separation}}) to avoid interfering with
other uses of similar functionality.

Protocols that instantiate multiple, independent instances of either
hash\_to\_curve or encode\_to\_curve MUST enforce domain separation
between those instances.
This requirement applies both in the case of multiple instances targeting
the same curve and in the case of multiple instances targeting different curves.
(This is because the internal hash\_to\_field primitive ({{hashtofield}})
requires domain separation to guarantee independent outputs.)

Domain separation is enforced with a domain separation tag (DST),

> which is a byte string constructed using the following requirements:

Reviewer: ```suggestion
which is a byte string constructed according to the following requirements:
```

- MUST be built on primitives designed for use in applications requiring
cryptographic randomness. As examples, a secure stream cipher is an appropriate
primitive, whereas a Mersenne twister pseudorandom number generator {{MT98}} is not.

- MUST NOT use rejection sampling.

- MUST give independent values for distinct (msg, DST, length) inputs.
Meeting this requirement is subtle.

> As a simplified example, hashing msg || DST does not work,

Reviewer: Does `||` need escaping as elsewhere?

Or maybe the escaping is unnecessary?

Author: It doesn't look like it (I see it render fine)?
